### ISSUE TITLE
Additional recompile occurring on Heroku with Elixir 1.15 causing boot issues

### ISSUE THREAD

**Issue by beerlington** [elixir-lang/elixir#12771]
### Elixir and Erlang/OTP versions

Erlang/OTP 25 [erts-13.2] [source] [64-bit] [smp:10:10] [ds:10:10:10] [async-threads:1] [jit]

Elixir 1.15.2 (compiled with Erlang/OTP 25)

### Operating system

Heroku-22

### Current behavior

We're using [Heroku Elixir buildpack](https://github.com/HashNuke/heroku-buildpack-elixir) to deploy and run our application on Heroku. After upgrading to Elixir 1.15, our application now does an additional recompile after the server restarts. With a large application like ours, the recompile takes longer than 30 seconds, causing Heroku to kill the process and the application will not boot.

I was able to reproduce this with a bare bones Phoenix application. You'll see in the output below where it's recompiling after the server starts.

```
2023-07-06T14:42:59.587968+00:00 heroku[web.1]: Restarting
2023-07-06T14:42:59.650735+00:00 heroku[web.1]: State changed from up to starting
2023-07-06T14:43:00.295043+00:00 heroku[web.1]: Stopping all processes with SIGTERM
2023-07-06T14:43:00.322078+00:00 app[web.1]: 14:43:00.321 [notice] SIGTERM received - shutting down
2023-07-06T14:43:00.322086+00:00 app[web.1]: 
2023-07-06T14:43:01.502645+00:00 heroku[web.1]: Process exited with status 0
2023-07-06T14:43:03.720077+00:00 heroku[web.1]: Starting process with command `mix phx.server`
2023-07-06T14:43:05.363804+00:00 app[web.1]: Compiling 15 files (.ex)
****
```

I [posted on the Elixir forum](https://elixirforum.com/t/heroku-app-recompiles-after-restarts-using-elixir-1-15/56893/3) about this issue and someone suggested trying mix releases. I tried that, but am still seeing an additional recompile happen that does not occur with 1.14. With mix releases, the recompile happens during Heroku's release phase, causing deploys to take a few extra minutes. This is better than the application not starting at all, but is still unexpected.

### Expected behavior

The application should not recompile after a restart on Heroku. With Elixir 1.14 this does not happen.

**Comment by josevalim** [comment]
We need a mechanism to reproduce this, otherwise there isn’t much we can do. Also, do you by any chance have `uniq` as a dependency?

**Comment by beerlington** [comment]
> We need a mechanism to reproduce this

Outside of Heroku? Or would you be able to test it with access to an app on Heroku?

> do you by any chance have uniq as a dependency?

No, our production application does not have this. It's also reproducible in a brand new Phoenix app, which does not appear to include this dep either.

**Comment by josevalim** [comment]
If it happens on Heroku, then it may be reproducible locally somehow? Maybe by running in prod and moving directories around? Do you see any dependency compile as well? Are all files compiled or only part of them?

**Comment by beerlington** [comment]
I'll try to reproduce it locally and will report back.

No dependencies are recompiled. It appears to be the entire application because it reports that it's compiling the same number of files as when I run ` mix compile --force` locally.

**Comment by beerlington** [comment]
Unfortunately I'm not able to reproduce it locally. I have limited knowledge of Heroku's filesystem and how they do restarts, but I'm wondering if it's something specific to their system that is causing it. 

We may switch to use mix releases since that will at least allow us to upgrade to 1.15 without affecting boot time after restarts. I'm curious if others have run into this and just haven't noticed because they're already using mix releases or because their application compiles in under 30 seconds.

I'm happy to provide any other details to try and get to the bottom of this though.

**Comment by beerlington** [comment]
I have one update to this. My original report said "With mix releases, the recompile happens during Heroku's release phase, causing deploys to take a few extra minutes. ". I was using `mix ecto.migrate` directly in our release script. Someone suggested following [this migration guide](https://hexdocs.pm/phoenix/releases.html#ecto-migrations-and-custom-commands) and that does in fact prevent the recompile during the release phase.

I think you can probably close this issue and I'll open another one on the elixir build pack project letting them know it only works on 1.15 when using mix releases.

**Comment by josevalim** [comment]
@beerlington deployment should work, even with Mix, but there is something else at play forcing the recompilation which we should address (if reproducible). I will keep this open in case someone else isolates this, and close in the future otherwise. Thanks!

**Comment by elliotb** [comment]
We're also seeing this on Heroku since upgrading to v1.15. If you have any suggestions regarding how we can isolate this issue, or e.g. trace what's triggering the compilation, then I'm happy to investigate too.

**Comment by josevalim** [comment]
The only obvious step I can think of is running https://github.com/heroku/builder to build a Docker image locally, so you can try to reproduce it.

**Comment by josevalim** [comment]
Folks, are any of you using path dependencies? (in an umbrella or not)?

**Comment by elliotb** [comment]
> Folks, are any of you using path dependencies? (in an umbrella or not)?

We're a non-Umbrella app, but whilst we don't have a path dependency, we do have a GitHub dependency - `{:kaffy, github: "aesmail/kaffy", branch: "master"},`.

**Comment by beerlington** [comment]
> Folks, are any of you using path dependencies? (in an umbrella or not)?

No path dependencies for my app either.

**Comment by moomerman** [comment]
Not 100% sure this is the same issue, but I've just noticed that our CI is now doing a full recompilation on every run since upgrading to 1.15 (and OTP 26) despite the cache being fully restored and no deps changes.

**Comment by josevalim** [comment]
Folks, any news on how to reproduce this issue? Thank you.

**Comment by beerlington** [comment]
@josevalim I was able to reproduce it using a brand new Phoenix app and deploying to Heroku with the [elixir buildpack](https://github.com/HashNuke/heroku-buildpack-elixir). I'm not sure if it's something unique to Heroku's environment or the buildpack, but I was not able to reproduce it locally.

We ended up switching to releases to get around the issue on Heroku, so I haven't investigated it any more.

**Comment by elliotb** [comment]
@josevalim we've also switched to releases, which has worked around the issue for us.

**Comment by josevalim** [comment]
Thanks folks, with someone can provide a minimal way to reproduce the error, I will gladly look into it. Otherwise I recommend using releases anyway, so we can go ahead and close this :) Glad to reopen once we can reproduce it directly.

**Comment by iamyojimbo** [comment]
Picking this back up. I was able to reproduce locally.

## TL;DR

Pass `--no-compile` once to all your mix commands in Heroku.

I.e, change your `Procfile` to:

```
release: mix do ecto.migrate --no-compile, any_other_example_mix_command
web: MIX_ENV=prod mix phx.server --no-compile
```

The behaviour is described here: https://github.com/elixir-lang/elixir/issues/13241

## Explanation

The issue is that Heroku performs the build in a tmp build dir, like
`/tmp/build_7c20b005`, but then copies the built code to the Dyno in `/app`. In
`_build/prod/lib/<<your_module>>/.mix/compile.elixir`, the cache key contains
the absolute path of where the source code was compiled, which is now
invalidated.

This is new in elixir v1.15. Here's the relevant commit:
<https://github.com/elixir-lang/elixir/commit/3699d94a33e5b65d3af8cec87b231e23566b4ad9>
found with: `git log v1.14.5..v1.15.0 --oneline --grep="cache key" -i`

This is an example diff of a formatted output of
`_build/prod/lib/elixir_heroku_recomp/.mix/compile.elixir` before and after
moving a project to a different dir and then recompiling it.

```diff
--- output/before_compile_elixir.txt 2025-08-15 14:29:47.453173063 +0100
+++ output/after_compile_elixir.txt 2025-08-15 14:29:47.452433725 +0100
@@ -47,12 +47,13 @@
 ================================================================================
 CACHE_KEY
 ================================================================================
-{[], ["lib"], "/Users/savvasnicholas/code/elixir-heroku-recomp/proj", false}
+{[], ["lib"], "/Users/savvasnicholas/code/elixir-heroku-recomp/tmp/proj_z1A9A5",
+ false}

 Cache key components:
   1. elixirc_options: []
   2. elixirc_paths (source directories): ["lib"]
-  3. working_directory (absolute path): "/Users/savvasnicholas/code/elixir-heroku-recomp/proj"
+  3. working_directory (absolute path): "/Users/savvasnicholas/code/elixir-heroku-recomp/tmp/proj_z1A9A5"
   4. no_optional_deps flag: false

 ================================================================================
```

I have created a repo here to allow you to reproduce the issue and the above
diff: <https://github.com/iamyojimbo/elixir-heroku-recomp>

**Comment by josevalim** [comment]
This is fantastic, thank you. I think we should start putting relative paths in source, especially because it will be helpful when working with git worktrees.

**Comment by josevalim** [comment]
Using relative paths came with other complications but we added a flag called `--no-check-cwd`, I recommend setting it. It will be in v1.19.2.

**Comment by sabiwara** [comment]
Linking the commit for the record: https://github.com/elixir-lang/elixir/commit/668fb69a4744fad4414d039689b2c648d3b091b6
